### PR TITLE
IR type mapping.

### PR DESCRIPTION
**Pull request by traceyyoshima**
TypeMapping from the PSI through the IR.

### CODE REVIEW

**Review by knutwannheden — COMMENTED** [review]
I have not looked at everything in detail yet, as there is quite a lot of changes and code, but I added a few small remarks.

Further, I think I noticed that the tests run considerably slower when I switch between the FIR-based and the PSI-based parser. Also when I disable a lot of debug logging, the `tree` tests take around 100s to run vs. 20s with the FIR-based parser. I have not compared this to the performance without this PR, but regardless of that I am a bit worried about the performance and I am unsure what the cause of this is (I have not done any profiling).

**Comment by kunli2** [comment]
> ba

good finding, Knut, this is something we do want to investigate and understand

**Comment by kunli2** [comment]
> I have not looked at everything in detail yet, as there is quite a lot of changes and code, but I added a few small remarks.
> 
> Further, I think I noticed that the tests run considerably slower when I switch between the FIR-based and the PSI-based parser. Also when I disable a lot of debug logging, the `tree` tests take around 100s to run vs. 20s with the FIR-based parser. I have not compared this to the performance without this PR, but regardless of that I am a bit worried about the performance and I am unsure what the cause of this is (I have not done any profiling).

I did some profiling, I think the 100s above include both PSI and FIR parsing, my test locally is about ~26 seconds (FIR-based-parsing) VS ~58 seconds (PSI-based-parsing).

File: src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt [hunk 1]
```diff
@@ -15,200 +15,247 @@
...
 import org.openrewrite.java.tree.JavaType
-import org.openrewrite.kotlin.KotlinTypeMapping
+import org.openrewrite.kotlin.KotlinIrTypeMapping
 
-class PsiElementAssociations(val typeMapping: KotlinTypeMapping, val file: FirFile) {
+class PsiElementAssociations(val typeMapping: KotlinIrTypeMapping, private val psiFile: PsiFile, val file: IrFile) {
 
-    private val elementMap: MutableMap<PsiElement, MutableList<FirInfo>> = HashMap()
+    private val notMapped: MutableList<IrElement> = ArrayList()
+    private val psiMap: MutableMap<TextRange, MutableList<PsiElement>> = HashMap()
```
[the remark this answers is not in the document]
Comment (traceyyoshima) [reply src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt:38]: > The fact that we need to use a text range to establish the mapping between some PSI and IR elements worries me a bit. Which cases exactly do we need this for? Would there maybe be some other way to achieve this (I am thinking by means of the "role" of the element within its parent)?

Yes, it is possible to link the IR to the PSI through the parent -- through `FirMetadataSource` elements: File, Script, Class, Function, and Property. But I planned to sort that out later. I'll look into that today instead.

File: src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt [hunk 2]
```diff
@@ -15,200 +15,247 @@
...
+                                elementMap.computeIfAbsent(psi.toList()[0]) { ArrayList() } += IrInfo(element, depth)
+                            } else {
+                                for (p in psi) {
+                                    elementMap.computeIfAbsent(p) { ArrayList() } += IrInfo(element, depth)
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    notMapped.add(element)
```
[the remark this answers is not in the document]
Comment (traceyyoshima) [reply src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt:106]: Yes, I want to be sure that nothing is missed. There are a lot of things that will be removed once the type mapping is finalized

**Comment by knutwannheden** [comment]
> I did some profiling, I think the 100s above include both PSI and FIR parsing, my test locally is about ~26 seconds (FIR-based-parsing) VS ~58 seconds (PSI-based-parsing).

@kunli2 Is that testing with the 58s on this branch?

File: src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (knutwannheden) [reply src/main/kotlin/org/openrewrite/kotlin/internal/PsiElementAssociations.kt:38]: @traceyyoshima That is fine. Knowing that we have an idea how to fix this is already good enough, I think.

**Review by kunli2 — APPROVED** [review]

**Comment by traceyyoshima** [comment]
> Further, I think I noticed that the tests run considerably slower when I switch between the FIR-based and the PSI-based parser. Also when I disable a lot of debug logging, the `tree` tests take around 100s to run vs. 20s with the FIR-based parser. I have not compared this to the performance without this PR, but regardless of that I am a bit worried about the performance and I am unsure what the cause of this is (I have not done any profiling).

A difference in type attribution likely causes the issue. I haven't done a thorough analysis yet, but in a simple case of an empty class, the FIR-based parser contains over 10k fewer types. Additionally, I hadn't checked for every instance where a signature is being unnecessarily recalculated.

**Comment by knutwannheden** [comment]
> I think it's reasonable that the IR-type-mapping takes longer as it generates more types.

We still need to keep an eye on this. I worry that it will be too slow to parse any sizeable repo. But we can hopefully improve this.

**Pull request merged**